RosNet/RosReader/RosBagReader.cs

```diff
@@ -2,67 +2,335 @@
+        byte[] unCompressedData = Array.Empty<byte>();
+        using MemoryStream source = new MemoryStream(data);
+        using MemoryStream target = new MemoryStream(compressedDataLength);
+        BZip2.Decompress(source, target, false);
+        unCompressedData = target.ToArray();
+
+        var connections = new List<Connection>();
+        using (BinaryReader tempReader = new BinaryReader(new MemoryStream(unCompressedData)))
+        {
+            connections = ReadChunk(tempReader, unParsedMessageHandler, unCompressedData.Length);
+        }
+
+        return connections;
+
+    }
+
+    /// <summary>
+    /// Reads chunk and adds all messages and message data to unParsedMessageHandler
+    /// </summary>
+    /// <returns>List of connections in chunk</returns>
+    private static List<Connection> ReadChunk(BinaryReader reader, UnParsedMessageHandler unParsedMessageHandler, int dataLength)
+    {
+        long endPos = reader.BaseStream.Position + dataLength;
+
+        var connections = new List<Connection>();
+
+        while (reader.BaseStream.Position != endPos)
+        {
+            Dictionary<string, FieldValue> header = ReadHeader(reader);
+
+            switch ((OpCode)header["op"].Value.First())
+            {
+                case OpCode.MessageData:
+                    var message = new Message(header["conn"], header["time"]);
+                    var data = ReadMessageData(reader);
+                    unParsedMessageHandler.AddUnParsedMessage(message, data);
+                    break;
+                case OpCode.Connection:
+                    var connection = new Connection(header["conn"], header["topic"]);
+                    SetConnectionData(reader, connection);
+                    connections.Add(connection);
+                    break;
+                default:
+                    var dataLen = reader.ReadInt32();
+                    reader.ReadBytes(dataLen);
+                    break;
             }
-            else
+        }
+        return connections;
+
+    }
+
+    /// <summary>
+    /// Reads a header
+    /// </summary>
+    /// <returns>Dictionary with fieldnames and fieldvalues in header</returns>
+    private static Dictionary<string, FieldValue> ReadHeader(BinaryReader reader)
+    {
+        int headerLen = reader.ReadInt32();
+        long headerEnd = reader.BaseStream.Position + headerLen;
+        var headerFields = new Dictionary<string, FieldValue>();
+
+        //checks if all fields in header are read
+        while (reader.BaseStream.Position < headerEnd)
+        {
+            FieldValue fieldValue = ReadField(reader);
+            headerFields.Add(fieldValue.Name, fieldValue);
+        }
+
+        if (!headerFields.ContainsKey("op"))
+            //TODO: More descriptive exception
+            throw new Exception("Header is missing op definition");
+
+        foreach (string headerField in HeaderFieldsByOp[(OpCode)headerFields["op"].Value.First()])
```

Comment:
```suggestion
        if (!HeaderFieldsByOp[(OpCode)op.Value.First()].All(h => headerFields.ContainsKey(h)))
```
(and change code later, can't suggest over deleted lines in GH-reviews)